pkg/controllers/management/nodepool/nodepool.go

```diff
@@ -56,77 +56,95 @@ func Register(ctx context.Context, management *config.ManagementContext) {
 
 	// Add handlers
 	p.NodePools.AddLifecycle(ctx, "nodepool-provisioner", p)
-	management.Management.Nodes("").AddHandler(ctx, "nodepool-provisioner", p.machineChanged)
+	management.Management.Nodes("").AddHandler(ctx, "nodepool-provisioner", p.nodeChanged)
 }
 
+// Create does a noop for nodepool.
 func (c *Controller) Create(nodePool *v3.NodePool) (runtime.Object, error) {
 	return nodePool, nil
 }
 
+// needsReconcile checks the nodepool to see if reconciliation is necessary.
 func (c *Controller) needsReconcile(nodePool *v3.NodePool, nodes []*v3.Node) bool {
 	changed, _, err := c.createOrCheckNodes(nodePool, nodes, true)
 	if err != nil {
-		logrus.Debugf("[nodepoool] error checking pool for reconciliation: %s", err)
+		logrus.Debugf("[nodepool] error checking pool for reconciliation: %s", err)
 	}
 
 	return changed
 }
 
-func (c *Controller) reconcile(nodePool *v3.NodePool, nodes []*v3.Node) {
+// reconcile reconciles the nodepool provided if necessary
+func (c *Controller) reconcile(nodePool *v3.NodePool, nodes []*v3.Node) (runtime.Object, error) {
 	_, qty, err := c.createOrCheckNodes(nodePool, nodes, false)
 	if err != nil {
-		logrus.Errorf("[nodepool] reconcile error, create or check nodes: %s", err)
+		return nodePool, fmt.Errorf("[nodepool] reconcile error, create or check nodes: %s", err)
 	}
 
+	// In certain cases, the nodepool quantity can actually decrease through a `createOrCheckNodes` call, like
+	// for scaling down a specific node. In this case, we need to update the nodePool object with the latest
+	// quantity so that on a future reconciliation loop, we do not recreate a node that was specifically scaled down.
 	if qty != nodePool.Spec.Quantity {
 		nodePool.Spec.Quantity = qty
 	}
 
-	_, err = c.setReconcileAnnotation(nodePool, "")
+	nodePool, err = c.updateNodePoolAnnotationAndSpec(nodePool, "", true)
 	if err != nil {
-		logrus.Errorf("[nodepool] error updating reconcile annotation to updated: %s", err)
+		return nodePool, fmt.Errorf("[nodepool] error updating reconcile annotation to updated: %s", err)
 	}
+	return nodePool, nil
 }
 
+// Updated reconciles a nodepool if necessary
 func (c *Controller) Updated(nodePool *v3.NodePool) (runtime.Object, error) {
+	if nodePool != nil {
+		logrus.Tracef("[nodepool] Updated called for nodepool %s: %+v", nodePool.Name, nodePool)
+	}
 	obj, err := v32.NodePoolConditionUpdated.Do(nodePool, func() (runtime.Object, error) {
```

Comment:
Do we rely on this condition to have the status/error msg? I think it won't reflect on the nodePool because there's no update on the `obj`

Reply:
marking resolved to keep old behavior as is for now